1. a="Hello Python" 문자열을 Slicing과 연결 연산자 (+)를 사용하여 "Python Hello"로 변경하시오.
    * [주의] a 변수 자체에 "Python Hello" 문자열이 저장되어야 함

In [46]:
a = "Hello Python"
a = a[6:] + ' ' + a[:6]
print(a)

Python Hello 


a의 6번째 인덱스부터 끝까지 슬라이싱("Python")한 후에 처음부터 5번째까지 슬라이싱("Hello")한 결과를 연결 연산자를 통해 이어붙인다.

즉, "Python Hello"가 된다.

1. b="Hello Python World" 문자열을 Slicing과 연결 연산자 (+)를 사용하여 "World Python Hello"로 변경하시오.
    * [주의] b 변수 자체에 "World Python Hello" 문자열이 저장되어야 함

In [50]:
b = "Hello Python World"
b = b[13:] + ' ' + b[6:13] + b[:6]
print(b)

World Python Hello 


b의 13번째 인덱스부터 끝까지 슬라이싱("World")한 후에 6번째부터 12번째까지 슬라이싱("Python")한 결과와 처음부터 5번째까지 슬라이싱("Hello")한 결과를 연결 연산자를 통해 이어붙인다.

즉, "World Python Hello"가 된다.

1. c="Hello"를 "olleH"로 변경하시오
    * [주의] c 변수 자체에 "olleH" 문자열이 저장되어야 함

In [51]:
c = "Hello"
c = c[::-1]
print(c)

olleH


c를 처음부터 끝까지 슬라이싱하는데 step이 -1이여서 뒤에서부터 슬라이싱된다.

즉, c를 뒤집은 것과 같이 "olleH"가 된다.

1. s="python"에 대해 다음 문제를 풀어보시오.
    1. s[0], s[0][0], s[0][0][0]은 각각 어떤 값이 나오는지 확인하고 그 이유를 나름대로 설명해 보시오.
    1. s[-100], s[100]은 값이 나오는지 에러가 나오는지 확인하고 그 결과에 대한 이유를 나름대로 설명해 보시오.
    1. s[-100, 100]은 값이 나오는지 에러가 나오는지 확인하고 그 결과에 대한 이유를 나름대로 설명해 보시오.
    1. s[1:-1]의 결과를 확인하고 그 결과에 대한 이유를 정확하게 설명하시오.
    1. s[3:-3]의 결과를 확인하고 그 결과에 대한 이유를 정확하게 설명하시오.

In [64]:
s = "python"
print(s[0], s[0][0], s[0][0][0])

p p p


첫 번째에서 s[0]은 s라는 레퍼런스에서 첫 번째 위치를 가져오라는 뜻이므로 당연히 첫 번째 글자인 'p'를 출력한다.

두 번째에서 s[0][0]은 s[0]라는 레퍼런스에서 첫 번째 위치를 가져오라는 뜻인데, s[0]은 위에서 말했듯이 'p'이다. 'p'의 첫 번째 위치는 'p'이므로 'p'를 출력한다.

세 번째에서 s[0][0][0]은 s[0][0]라는 레퍼런스에서 첫 번째 위치를 가져오라는 뜻이다. 두 번째와 동일하게 s[0][0]은 'p'이므로 'p'의 첫 번째 위치인 'p'를 출력한다.

In [55]:
s = "python"
print(s[-100], s[100])

IndexError: string index out of range

s[-100]과 s[100]의 의미는 뒤에서부터 인덱스가 100인 부분과 앞에서부터 인덱스가 100인 부분을 반환하는 것이다.

하지만 s는 "python"으로 길이가 6인 문자열이다.

그러므로 뒤에서부터 100인 부분과 앞에서부터 100인 부분은 존재하지 않아 범위를 벗어났다는 오류를 출력한다.

In [54]:
s = "python"
print(s[-100, 100])

TypeError: string indices must be integers

문자열 슬라이싱의 형식은 [start(included) : stop(excluded) : step]으로 start와 stop, step은 모두 int형 정수여야 한다.

하지만 -100, 100은 List의 형태이기 때문에 integer이 들어와야 한다는 오류를 출력한다.

In [57]:
s = "python"
print(s[1:-1])

ytho


시작하는 인덱스가 1이고 (excluded 된) 마지막 인덱스가 -1이다.

s에서 인덱스 1의 부분은 'y'이고 -1의 부분은 'n', 즉 'y'부터 'n' 이전까지라는 의미가 된다.

즉, "ytho"가 된다.

In [62]:
s = "python"
print(s[3:-3])

시작하는 인덱스가 3이고 (excluded 된) 마지막 인덱스가 -3이다.

s에서 인덱스 3의 부분은 'h'이고 -3의 부분은 '-3', 즉 'h'부터 'h' 이전까지라는 의미인데 두 부분이 똑같다.

그러므로 아무것도 슬라이싱되지 않는다.

1. for문을 활용하여 1부터 100사이의 홀수를 출력하시오.

In [69]:
for i in range(1, 100, 2): print(i, end=', ')

1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99, 

1부터 100까지 2씩 증가시켜 가며 출력한다. 이는 결국 홀수들만 출력하는 것과 마찬가지이다.

1. while문을 활용하여 1부터 100사이의 짝수의 합을 계산하여 출력하시오.

In [3]:
a = 0
i = 0
while i < 101:
    if i % 2 == 0: a += i
    i += 1
print(a)

2550


i가 101을 넘어가기 전까지 짝수 여부를 판별하여 a에 더한다. a에는 1에서 100 사이의 짝수들의 합이 담긴다.

1. 사용자로 부터 임의의 정수를 입력받고, 해당 숫자를 역순으로 출력하는 프로그램을 작성하시오.

In [70]:
a = int(input('정수를 입력하세요: '))
b = 0
print('숫자 역순은 ', end='')
while a > 0:
    b = b * 10 + a % 10
    a = int(a / 10)
print(b, end='')
print(' 입니다.')

정수를 입력하세요: 123
숫자 역순은 321 입니다.


a를 10으로 나눈 나머지 값들을 b에 10을 곱한 후 더한다. 그리고 a는 a를 10으로 나눈 몫으로 바꾼다.

이를 예로 들어 설명하면,

123이라는 숫자가 있을 때, 10으로 나눈 나머지는 3이므로 b는 0 x 10 + 3으로 3이 되고 a는 10으로 나눈 몫인 12가 된다.
다음엔 12를 10으로 나눈 나머지 2, b는 3 x 10 + 2가 되어 32이고 a는 1이다.
마지막으로 나머지 1, b는 32 x 10 + 1로 321이고 a는 0이다.

결국 b는 a를 뒤집은 숫자가 된다.

1. 사용자로 부터 정수를 입력받아서 1부터 그 사이에 존재하는 소수 (Prime number)를 출력하는 파이썬 프로그램을 작성하시오.

In [60]:
a = int(input())
pn = []
for i in range(0, a + 1): pn.append(1)
for i in range(2, a + 1):
    if pn[i] == 0: continue
    for j in range(i * i, a + 1, i): pn[j] = 0
for i in range(2, a + 1):
    if pn[i]: print(i, end=',')

20
2,3,5,7,11,13,17,19,

에라토스테네스의 체를 이용하여 소수들을 구하였다.

1. [ACM-ICPC 문제 2441번]
    * 별찍기-4

In [38]:
a = int(input())
for i in range(a, 0, -1):
    for j in range(0, a - i):
        print(' ', end='')
    for j in range(0, i):
        print('*', end='')
    print()

5
*****
 ****
  ***
   **
    *


N번째 줄에는 (N - 1)개의 공백을 출력하고 나머지 칸에 별을 출력하면 된다.

![](https://raw.githubusercontent.com/knight7024/Script-Programming-2018-3/master/09-20-1.PNG "1번")

1. [ACM-ICPC 문제 11721번] 
    * 열 개씩 끊어 출력하기

In [71]:
a = input()
for i in range(len(a)):
    print(a[i], end='')
    if (i+1) % 10 == 0:print()

OneTwoThreeFourFiveSixSevenEightNineTen
OneTwoThre
eFourFiveS
ixSevenEig
htNineTen

입력받은 문자열의 길이만큼 반복문을 돌다가 현재 위치가 10번째의 배수인지 아닌지를 판단하여 공백을 출력한다.

![](https://raw.githubusercontent.com/knight7024/Script-Programming-2018-3/master/09-20-2.PNG "2번")

1. [ACM-ICPC 문제 2839번]
    * 설탕배달

In [41]:
a = int(input())
t = 0
f = 0
while a > 0 and a % 5:
    a = a - 3
    t = t + 1
f = int(a / 5)
if a < 0: print(-1)
else: print(t + f)

18
4


그리디하게 생각하면 5g의 설탕이 가장 많으면 된다. 그러므로 입력받은 숫자에서 3만큼 빼다가 5의 배수가 된다면 반복문을 종료하면 된다.

음수가 된다면 3과 5로도 표현할 수 없는 숫자이므로 -1을 출력하고 아니면 구한 결과를 출력한다.

![](https://raw.githubusercontent.com/knight7024/Script-Programming-2018-3/master/09-20-3.PNG "3번")

### 과제 소감 :
평소 사용해오던 C언어, 심지어 스크립트 언어인 Ruby와도 문법이 다른 부분이 많고 '모든 것을 레퍼런스 취급하지만 Mutable한 자료형은 따로 있다' 같은 파이썬만의 고유한 특징들 때문에 사용하는데 어려움을 많이 겪었다.

다만 모든 구문의 길이가 짧다는 특징때문에 편리하게 사용할 수 있었다.